content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -1051,25 +1051,58 @@ A portion of the schema is correlatable if all `array` parent schemas are of typ
 
 #### Resource Constraints
 
-CEL expressions have the potential to consume unacceptable amounts of API server resources. We constrain the resource utilization in following ways:
-- Validation of CEL expression's "cost" when a CEL expression is written to a field in a CRD (at CRD creation/update time)
-- Runtime cost budget during CEL evaluation
-  - CEL validation might fail due to runtime cost budget exceed with error message `validation failed due to running out of cost budget, no further validation rules will be run`
-  - CEL validation might fail due to cost limit exceed per expression with message `operation cancelled: actual cost limit exceeded: no further validation rules will be run due to call cost exceeds limit for rule:{$rule}`
-- Go context cancellation to bound CEL expression evaluation to the request lifetime
-
-Guidelines for working with estimated limits:
-- Adding MaxItems, MaxProperties and MaxLength limits on all data accessed by CEL rules is the best practice.
-- O(n) - For simple rules, it is possible to iterate across a single map/list/string without exceeding the limit, but adding limits on all data accessed by CEL rules is the best practice
-- O(n^2)+ the product of the max lengths usually needs to be <1,000,000.  E.g. 1000 for 2 levels of nesting, 100 for 3 levels of nesting
-- O(n^3) - should generally be avoided
-
-Remember the big-O of a rule takes into account how many times the rule will be executed in addition to
-the data that the rule is working with. If a list inside of a list has a CEL expression that uses `self.all`,
-that rule is O(mn), not O(n). Rules that operate on nested aggregate types (lists inside maps, maps inside 
-maps, and so on) will  likely need to have appropriate limits set on all lists/maps involved before cost 
-validation will pass. Rules that perform string operations or use regular expressions inside of a list or map 
-will also likely need to have limits set in as many places as possible for similar reasons.
+Resource consumption of CEL expressions is limited at compile time by static analysis. If an expression is
```

Comment:
It may not be obvious to readers what "static analysis" means here. Explain in terms of the CRD operations that trigger the cost check?